### ISSUE TITLE
iam_role: permissions_boundary attribute should accept empty string as absent value

### ISSUE THREAD

**Issue by mgarstecki** [terraform-providers/terraform-provider-aws#5546]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```
Terraform v0.11.7
+ provider.aws v1.31.0
+ provider.template v1.0.0
```

### Affected Resource(s)

* aws_iam_role

### Terraform Configuration Files

```hcl
resource "aws_iam_role" "test_role" {
  name = "test_role"
  assume_role_policy = "${data.aws_iam_policy_document.assume_role.json}"
  permissions_boundary = ""
}

data "aws_iam_policy_document" "assume_role" {
  statement {
    actions = ["sts:AssumeRole"]
    effect  = "Allow"

    principals {
      type        = "Service"
      identifiers = ["ec2.amazonaws.com"]
    }
  }
}
```

### Debug Output

Debug log of `terraform plan` : https://gist.github.com/mgarstecki/62cec1b01846e74afbedab1589581c1e

### Expected Behavior

Terraform should generate an execution plan to create the role without a permission boundary.

### Actual Behavior

Terraform fails with the following error :

```
Error: aws_iam_role.test_role: expected length of permissions_boundary to be in the range (20 - 2048), got 
```
### Steps to Reproduce

1. `terraform plan` or `terraform apply`

### Important Factoids

The `permissions_boundary` attribute is optional, and it seems common in TF to treat an empty value as absent for the sake of validation. It looks more like a bug than a feature request to me, but feel free to correct me if this is not how the community would classify this.

Tolerating the empty value is useful in my context : we maintain a module that creates roles, and should work for users with and without a permission boundary constraint. I want to have a `permissions_boundary` input variable to handle both cases:

```hcl
variable "permissions_boundary_arn" {
  default = ""
}

resource "aws_iam_role" "role" {
  name_prefix = "delegated-instance_"

  assume_role_policy = "..."
  permissions_boundary = "${var.permissions_boundary_arn}"
}
```

The current validation rule would oblige me to duplicate the module, one version for users without boundaries, and one version for those that use one.

I'll work on a PR to correct this.

**Comment by bflad** [comment]
The configuration language of Terraform (HCL) currently has some limitations when it comes to typing and lacks the ability to pass an "empty"/`null` value. The good news is that this is likely being addressed in the next major version of Terraform: https://www.hashicorp.com/blog/terraform-0-12-conditional-operator-improvements

**Comment by mgarstecki** [comment]
Agreed @bflad, the `null` value of 0.12 will solve this.

Still, tolerating an empty string would still be nice: it fits with how many optional values are handled elsewhere in the provider, making for a better UX I think. In a more selfish reason, it will also allow the above pattern to work in our modules while we wait for 0.12, and while we implement the sizable migration required in our large TF deployment...

So I still plan to work on a PR by the end of the week to allow the empty value, would you consider merging it even though 0.12 is coming up ?

Regards,

Mathieu G.

**Comment by bflad** [comment]
This has been released in [version 1.35.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1350-september-06-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
